tools/cuda_process_stream.cu

```diff
@@ -20,187 +78,119 @@ int main(int argc, char **argv) {
   int device_count = 0;
   cudaGetDeviceCount(&device_count);
   std::cout << "CUDA Device Count: " << device_count << "\n";
   std::cout << "CUDA Device ID: " << device_id << "\n";
 
-  int maxBytes = num_elems * num_sketches * sizeof(vec_t_cu) + num_elems * num_sketches * sizeof(vec_hash_t);
-  cudaGraph.cudaKernel.kernelUpdateSharedMemory(maxBytes);
+  size_t maxBytes = num_buckets * sizeof(vec_t_cu) + num_buckets * sizeof(vec_hash_t);
+  cc_gpu_alg.updateKernelSharedMemory(maxBytes);
   std::cout << "Allocated Shared Memory of: " << maxBytes << "\n";
 
   // Prefetch memory to device 
-  gpuErrchk(cudaMemPrefetchAsync(sketchSeeds, num_nodes * num_sketches * sizeof(long), device_id));
-
-  /*gpuErrchk(cudaMemPrefetchAsync(cudaSketches, num_nodes * num_sketches * sizeof(CudaSketch), device_id));
-  gpuErrchk(cudaMemPrefetchAsync(d_bucket_a, num_nodes * num_sketches * num_elems * sizeof(vec_t), device_id));
-  gpuErrchk(cudaMemPrefetchAsync(d_bucket_c, num_nodes * num_sketches * num_elems * sizeof(vec_hash_t), device_id));*/
-
-  cudaGraph.configure(&cudaUpdateParams, supernodes, sketchSeeds, num_threads);
+  gpuErrchk(cudaMemPrefetchAsync(sketchSeeds, num_nodes * sizeof(long), device_id));
+  cc_gpu_alg.configure(&cudaUpdateParams, sketchSeeds, num_threads);
   
-  MT_StreamReader reader(stream);
-  GraphUpdate upd;
-
   std::cout << "Finished initializing CUDA parameters\n";
   std::chrono::duration<double> init_time = std::chrono::steady_clock::now() - init_start;
   std::cout << "CUDA parameters init duration: " << init_time.count() << std::endl;
 
   // Start timer for kernel
   auto ins_start = std::chrono::steady_clock::now();
+  std::thread querier(track_insertions, num_updates, &driver, ins_start);
 
   // Call kernel code
   std::cout << "Update Kernel Starting...\n";
 
-  // Do the edge updates
-  std::vector<std::thread> threads;
-  threads.reserve(reader_threads);
-  auto task = [&](const int thr_id) {
-    MT_StreamReader reader(stream);
-    GraphUpdate upd;
-    while(true) {
-      upd = reader.get_edge();
-      if (upd.type == BREAKPOINT) break;
-      Edge &edge = upd.edge;
-
-      gts->insert({edge.src, edge.dst}, thr_id);
-      std::swap(edge.src, edge.dst);
-      gts->insert({edge.src, edge.dst}, thr_id);
-    }
-  };
+  // Perform edge updates
+  driver.process_stream_until(END_OF_STREAM);
 
-  // start inserters
-  for (int t = 0; t < reader_threads; t++) {
-    threads.emplace_back(task, t);
-  }
-  // wait for inserters to be done
-  for (int t = 0; t < reader_threads; t++) {
-    threads[t].join();
-  }
+  std::cout << "Update Kernel finished.\n";
 
-  std::cout << "  Flush Starting...\n";
+  // Start timer for cc
+  auto cc_start = std::chrono::steady_clock::now();
   
+  std::cout << "  Flush Starting...\n";
+
   auto flush_start = std::chrono::steady_clock::now();
+
   gts->force_flush();
-  GraphWorker::pause_workers();
+  driver.get_worker_threads()->flush_workers();
```

Comment:
Suggest calling `driver.prep_query()`

Reply:
Applied the suggestion in this commit: https://github.com/MangoShip/GraphZeppelin/pull/3/commits/64318f733ecf117aef6313d12743606adfc82542#diff-8353bfcb2001e1f9c38c566731763e6dccf375adb3d3e5a52b10c0bbc8fdc50dR100